{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G G G A A G C T T C C T\n",
      "12\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "def form_profile(motifs):\n",
    "    '''For each position in motifs, calculate the frequencies of ACGT'''\n",
    "    k = len(motifs[0])\n",
    "    t = len(motifs)\n",
    "    profile = {i: [0] * k for i in range(len('ACGT'))}\n",
    "\n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        for j, nucleotide in enumerate('ACGT'):\n",
    "            profile[j][i] = (column.count(nucleotide) + 1)/ t\n",
    "    return profile\n",
    "\n",
    "\n",
    "def Score(motifs):\n",
    "    '''Finds a consensus from motifs and a score indicating how well motifs match it'''\n",
    "    consensus = ''\n",
    "    score = 0\n",
    "    k = len(motifs[0])\n",
    "    counts = np.zeros((4, k))\n",
    "    \n",
    "    for i in range(k):\n",
    "        column = [motif[i] for motif in motifs]\n",
    "        counts[0][i] = column.count('A')\n",
    "        counts[1][i] = column.count('C')\n",
    "        counts[2][i] = column.count('G')\n",
    "        counts[3][i] = column.count('T')\n",
    "        \n",
    "        max_count = max(counts[:, i])\n",
    "        score += sum(counts[:, i]) - max_count\n",
    "        \n",
    "        if max_count == counts[0][i]:\n",
    "            consensus += 'A'\n",
    "        elif max_count == counts[1][i]:\n",
    "            consensus += 'C'\n",
    "        elif max_count == counts[2][i]:\n",
    "            consensus += 'G'\n",
    "        elif max_count == counts[3][i]:\n",
    "            consensus += 'T'\n",
    "    \n",
    "    return score, consensus\n",
    "\n",
    "\n",
    "def GreedyMotifFinding(seq, k, profile):\n",
    "    '''Finds the most probable k-mer in a sequence given a profile'''\n",
    "    max_prob = -1\n",
    "    kmer = ''\n",
    "    for i in range(len(seq) - k + 1):\n",
    "        kmer = seq[i:i+k]\n",
    "        prob = 1\n",
    "        for j in range(k):\n",
    "            if kmer[j] == 'A':\n",
    "                prob *= profile[0][j]\n",
    "            elif kmer[j] == 'C':\n",
    "                prob *= profile[1][j]\n",
    "            elif kmer[j] == 'G':\n",
    "                prob *= profile[2][j]\n",
    "            elif kmer[j] == 'T':\n",
    "                prob *= profile[3][j]\n",
    "        \n",
    "        if prob > max_prob:\n",
    "            max_prob = prob\n",
    "            most_probable_kmer = kmer\n",
    "            \n",
    "    return most_probable_kmer\n",
    "\n",
    "def greedysearch(dna, k , t):\n",
    "    '''For a given k-mer length and given number of sequences given by t, \n",
    "    implements a Greedy Search algorithm to find the best motifs'''\n",
    "    bestmotifs = [seq[:k] for seq in dna]\n",
    "    for i in range(len(dna[0]) - k + 1):\n",
    "        motifs_first = [dna[0][i:i+k]]\n",
    "        for motif in motifs_first:\n",
    "            motifs_it = [motif]\n",
    "            for j in range(1, t):\n",
    "                profile = form_profile(motifs_it)\n",
    "                most = GreedyMotifFinding(dna[j], k, profile)\n",
    "                motifs_it.append(most)\n",
    "            if Score(motifs_it)[0] < Score(bestmotifs)[0]:\n",
    "                bestmotifs = motifs_it\n",
    "                consensus = Score(bestmotifs)[1]\n",
    "    return consensus\n",
    "   \n",
    "with open(\"dataset_30306_9.txt\", 'r') as file:\n",
    "    first_line = file.readline().strip().split()\n",
    "    k, t = map(int, first_line)\n",
    "    dna = file.read().strip().split()\n",
    "print(' '.join(greedysearch(dna, k , t)))\n",
    "print(len(greedysearch(dna, k, t)))\n",
    "# dna = ['GGCGTTCAGGCA', 'AAGAATCAGTCA', 'CAAGGAGTTCGC', 'CACGTCAATCAC', 'CAATAATATTCG']\n",
    "# k, t = 3, 5\n",
    "# print(' '.join(greedysearch(dna, k, t)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTAGGAA\n"
     ]
    }
   ],
   "source": [
    "def Neighbors(pattern, d):\n",
    "    '''Generate all sequences that differ from a pattern by max d nucleotides'''\n",
    "    if d == 0:\n",
    "        return pattern\n",
    "    if len(pattern) == 1:\n",
    "        return {'A', 'C', 'G', 'T'}\n",
    "    alphabet = {'A', 'C', 'G', 'T'}\n",
    "    neighbors = set()\n",
    "    \n",
    "    suffixneighbors = Neighbors(pattern[1:],d)\n",
    "    for text in suffixneighbors:\n",
    "        if HammingDistance(pattern[1:], text) < d:\n",
    "            for nucleotide in alphabet:\n",
    "                neighbors.add(nucleotide+text)\n",
    "        else:\n",
    "            neighbors.add(pattern[0] + text)\n",
    "    return neighbors\n",
    "\n",
    "def allstrings_try(k):\n",
    "    '''Generates all possible 4^k k-mers using Neighbors function'''\n",
    "    pattern = 'A' * k\n",
    "    seq = list(Neighbors(pattern, k))\n",
    "    return seq\n",
    "\n",
    "def DistanceBetweenPatternAndStrings(pattern, dna):\n",
    "    '''For a given pattern, compute its best match (minimum Hamming distance) in each dna string\n",
    "    dna -> a list of dna strings\n",
    "    pattern -> pattern'''\n",
    "    k = len(pattern)\n",
    "    distance = 0\n",
    "    for seq in dna:\n",
    "        hamming = float('inf')\n",
    "        for i in range(len(seq) - k + 1):\n",
    "            pattern_prime = seq[i:i+k]\n",
    "            if hamming >= HammingDistance(pattern_prime, pattern):\n",
    "                hamming = HammingDistance(pattern_prime, pattern)\n",
    "        distance += hamming\n",
    "    return distance\n",
    "\n",
    "def HammingDistance(seq1, seq2):\n",
    "    count = 0\n",
    "    for n, j in zip(seq1, seq2):\n",
    "        if n != j:\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "def MedianString(dna, k):\n",
    "    '''Finds the k-mer that has the minimum total Hamming distance to all DNA strings\n",
    "    dna -> list of dna strings\n",
    "    k -> length of the k-mer'''\n",
    "    d = float('inf')\n",
    "    patterns = allstrings_try(k)\n",
    "    median = ''\n",
    "    for pattern in patterns:\n",
    "        if DistanceBetweenPatternAndStrings(pattern, dna) <= d:\n",
    "            d = DistanceBetweenPatternAndStrings(pattern, dna)\n",
    "            median = pattern\n",
    "    return median\n",
    "\n",
    "\n",
    "\n",
    "k = 7\n",
    "dna = ['CTCGATGAGTAGGAAAGTAGTTTCACTGGGCGAACCACCCCGGCGCTAATCCTAGTGCCC', 'GCAATCCTACCCGAGGCCACATATCAGTAGGAACTAGAACCACCACGGGTGGCTAGTTTC', 'GGTGTTGAACCACGGGGTTAGTTTCATCTATTGTAGGAATCGGCTTCAAATCCTACACAG']\n",
    "\n",
    "print(MedianString(dna, k))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
